### PR TITLE
Add unit tests for PAP Status Data Handler classes

### PR DESCRIPTION
**Pull request by UdeshAthukorala**
### Proposed changes in this pull request

$Subject

### CODE REVIEW

File: components/entitlement/org.wso2.carbon.identity.entitlement/src/test/java/org/wso2/carbon/identity/entitlement/persistence/PAPStatusDataHandlerTest.java [hunk 1]
```diff
@@ -0,0 +1,330 @@
...
+
+    StatusHolder statusHolderForAddPolicy;
+    StatusHolder statusHolderForGetPolicy;
+    StatusHolder statusHolderForRemovePolicy;
+
+    StatusHolder statusHolderForAddPolicyInPDP;
+    StatusHolder statusHolderForUpdatePolicyInPDP;
+    StatusHolder statusHolderForRemovePolicyInPDP;
+
+    List<StatusHolder> policyStatusHolderList1;
```
Comment (darshanasbg) [inline components/entitlement/org.wso2.carbon.identity.entitlement/src/test/java/org/wso2/carbon/identity/entitlement/persistence/PAPStatusDataHandlerTest.java:77]: Shall we use more appropriate variable names that indicate the usage of it?

File: components/entitlement/org.wso2.carbon.identity.entitlement/src/test/java/org/wso2/carbon/identity/entitlement/persistence/PAPStatusDataHandlerTest.java [hunk 2]
```diff
@@ -0,0 +1,330 @@
...
+        setupPolicyStatusData();
+        setupSubscriberStatusData();
+    }
+
+    @AfterMethod
+    public void tearDown() throws Exception {
+
+        papStatusDataHandler.handle(ABOUT_POLICY, POLICY_KEY, policyStatusHolderList3);
+        papStatusDataHandler.handle(ABOUT_SUBSCRIBER, SUBSCRIBER_KEY, pdpStatusHolderList3);
+        subscriberPersistenceManager.removeSubscriber(SUBSCRIBER_KEY);
```
Comment (darshanasbg) [inline components/entitlement/org.wso2.carbon.identity.entitlement/src/test/java/org/wso2/carbon/identity/entitlement/persistence/PAPStatusDataHandlerTest.java:112]: what exactly we do here? is it about resetting data to a base level?

if so, its bit odd to use the methods that we want to test, to do the resetting.. If there is an issue in the `handle()` method or `remove()` method, many other tests will also fails.. making the identifying actual root cause for such failure will be harder due to the unnecessary noise..

File: components/entitlement/org.wso2.carbon.identity.entitlement/src/test/java/org/wso2/carbon/identity/entitlement/persistence/PAPStatusDataHandlerTest.java [hunk 3]
```diff
@@ -0,0 +1,330 @@
...
+
+    @Test(priority = 1)
+    public void testHandlePolicyStatus() throws Exception {
+
+        papStatusDataHandler.handle(ABOUT_POLICY, POLICY_KEY, policyStatusHolderList1);
+
+        StatusHolder[] statusDataFromStorage =
+                papStatusDataHandler.getStatusData(ABOUT_POLICY, POLICY_KEY, ADD_POLICY, "*");
+        assertEquals(statusDataFromStorage.length, 1);
+        assertEquals(statusDataFromStorage[0].getType(), statusHolderForAddPolicy.getType());
```
Comment (darshanasbg) [inline components/entitlement/org.wso2.carbon.identity.entitlement/src/test/java/org/wso2/carbon/identity/entitlement/persistence/PAPStatusDataHandlerTest.java:123]: here we add `policyStatusHolderList1` but validate the value against a different object `statusHolderForAddPolicy`.. one should have to have the understanding of the whole test class to correlate whats the relation between `policyStatusHolderList1` & `statusHolderForAddPolicy`

File: components/entitlement/org.wso2.carbon.identity.entitlement/src/test/java/org/wso2/carbon/identity/entitlement/persistence/PAPStatusDataHandlerTest.java [hunk 4]
```diff
@@ -0,0 +1,330 @@
...
+            papStatusDataHandler.handle(ABOUT_POLICY, POLICY_KEY, policyStatusHolderList1);
+            StatusHolder[] statusDataFromStorage =
+                    papStatusDataHandler.getStatusData(ABOUT_POLICY, POLICY_KEY, null, "*");
+            assertEquals(statusDataFromStorage.length, 1);
+
+            papStatusDataHandler.handle(ABOUT_POLICY, POLICY_KEY, policyStatusHolderList2);
+            statusDataFromStorage = papStatusDataHandler.getStatusData(ABOUT_POLICY, POLICY_KEY, null, "*");
+            assertEquals(statusDataFromStorage.length, 1);
+
+            assertEquals(statusDataFromStorage[0].getType(), statusHolderForGetPolicy.getType());
```
Comment (darshanasbg) [inline components/entitlement/org.wso2.carbon.identity.entitlement/src/test/java/org/wso2/carbon/identity/entitlement/persistence/PAPStatusDataHandlerTest.java:153]: same comment [1], applicable here as well.

[1] https://github.com/wso2/carbon-identity-framework/pull/5949/files#r1772661724

File: components/entitlement/org.wso2.carbon.identity.entitlement/src/test/java/org/wso2/carbon/identity/entitlement/persistence/PAPStatusDataHandlerTest.java [hunk 5]
```diff
@@ -0,0 +1,330 @@
...
+
+    @Test(priority = 5)
+    public void testHandleSubscriberStatus() throws Exception {
+
+        papStatusDataHandler.handle(ABOUT_SUBSCRIBER, SUBSCRIBER_KEY, pdpStatusHolderList1);
+
+        StatusHolder[] statusDataFromStorage =
+                papStatusDataHandler.getStatusData(ABOUT_SUBSCRIBER, SUBSCRIBER_KEY, null, "*");
+        assertEquals(statusDataFromStorage.length, 1);
+        assertEquals(statusDataFromStorage[0].getType(), statusHolderForAddPolicyInPDP.getType());
```
Comment (darshanasbg) [inline components/entitlement/org.wso2.carbon.identity.entitlement/src/test/java/org/wso2/carbon/identity/entitlement/persistence/PAPStatusDataHandlerTest.java:197]: same comment [1], applicable here as well.

[1] https://github.com/wso2/carbon-identity-framework/pull/5949/files#r1772661724

File: components/entitlement/org.wso2.carbon.identity.entitlement/src/test/java/org/wso2/carbon/identity/entitlement/persistence/HybridPAPStatusDataHandlerTest.java [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+    public void testGetStatusWhenPolicyStatusExistsInDb() throws Exception {
+
+        jdbcSimplePAPStatusDataHandler.handle(ABOUT_POLICY, POLICY_KEY, policyStatusHolderList1);
+        verifyStatusDataFromStorage();
+    }
+
+    @Test(priority = 13)
+    public void testGetStatusWhenPolicyStatusExistsInRegistry() throws Exception {
+
+        jdbcSimplePAPStatusDataHandler.handle(ABOUT_POLICY, POLICY_KEY, policyStatusHolderList1);
```
Comment (darshanasbg) [inline components/entitlement/org.wso2.carbon.identity.entitlement/src/test/java/org/wso2/carbon/identity/entitlement/persistence/HybridPAPStatusDataHandlerTest.java:139]: Should this be `registrySimplePAPStatusDataHandler`?

**Review by darshanasbg — CHANGES_REQUESTED** [review]

File: components/entitlement/org.wso2.carbon.identity.entitlement/src/test/java/org/wso2/carbon/identity/entitlement/persistence/PAPStatusDataHandlerTest.java (hunk 2 above)
Comment (UdeshAthukorala) [reply components/entitlement/org.wso2.carbon.identity.entitlement/src/test/java/org/wso2/carbon/identity/entitlement/persistence/PAPStatusDataHandlerTest.java:112]: Yes. This is about resetting data for empty state. There is no any other method to remove PAP Status related data. This status data removing part also done from the `handle()` method based on the Status type `DELETE_POLICY`.

File: components/entitlement/org.wso2.carbon.identity.entitlement/src/test/java/org/wso2/carbon/identity/entitlement/persistence/HybridPAPStatusDataHandlerTest.java (hunk 6 above)
Comment (UdeshAthukorala) [reply components/entitlement/org.wso2.carbon.identity.entitlement/src/test/java/org/wso2/carbon/identity/entitlement/persistence/HybridPAPStatusDataHandlerTest.java]: Adressed with https://github.com/wso2/carbon-identity-framework/compare/e393b967246d88953a960b77a0c4a7c355fd9dfd..f26be27fb80abcc3adb9aace8db0909c0eb3c17d

File: components/entitlement/org.wso2.carbon.identity.entitlement/src/test/java/org/wso2/carbon/identity/entitlement/persistence/PAPStatusDataHandlerTest.java [hunk 7]
```diff
@@ -0,0 +1,330 @@
...
+        assertEquals(statusDataFromStorage[0].getKey(), statusHolderForAddPolicyInPDP.getKey());
+        assertEquals(statusDataFromStorage[0].getTarget(), statusHolderForAddPolicyInPDP.getTarget());
+        assertEquals(statusDataFromStorage[0].getTargetAction(), statusHolderForAddPolicyInPDP.getTargetAction());
+        assertEquals(statusDataFromStorage[0].getUser(), statusHolderForAddPolicyInPDP.getUser());
+
+        papStatusDataHandler.handle(ABOUT_SUBSCRIBER, SUBSCRIBER_KEY, pdpStatusHolderList2);
+        statusDataFromStorage = papStatusDataHandler.getStatusData(ABOUT_SUBSCRIBER, SUBSCRIBER_KEY, null, "*");
+        assertEquals(statusDataFromStorage.length, 2);
```
[the remark this answers is not in the document]
Comment (UdeshAthukorala) [reply components/entitlement/org.wso2.carbon.identity.entitlement/src/test/java/org/wso2/carbon/identity/entitlement/persistence/PAPStatusDataHandlerTest.java]: Addressed with https://github.com/wso2/carbon-identity-framework/compare/e393b967246d88953a960b77a0c4a7c355fd9dfd..f26be27fb80abcc3adb9aace8db0909c0eb3c17d

File: components/entitlement/org.wso2.carbon.identity.entitlement/src/test/java/org/wso2/carbon/identity/entitlement/persistence/PAPStatusDataHandlerTest.java (hunk 5 above)
Comment (UdeshAthukorala) [reply components/entitlement/org.wso2.carbon.identity.entitlement/src/test/java/org/wso2/carbon/identity/entitlement/persistence/PAPStatusDataHandlerTest.java]: Addressed with https://github.com/wso2/carbon-identity-framework/compare/e393b967246d88953a960b77a0c4a7c355fd9dfd..f26be27fb80abcc3adb9aace8db0909c0eb3c17d

File: components/entitlement/org.wso2.carbon.identity.entitlement/src/test/java/org/wso2/carbon/identity/entitlement/persistence/PAPStatusDataHandlerTest.java (hunk 4 above)
Comment (UdeshAthukorala) [reply components/entitlement/org.wso2.carbon.identity.entitlement/src/test/java/org/wso2/carbon/identity/entitlement/persistence/PAPStatusDataHandlerTest.java]: Addressed with https://github.com/wso2/carbon-identity-framework/compare/e393b967246d88953a960b77a0c4a7c355fd9dfd..f26be27fb80abcc3adb9aace8db0909c0eb3c17d

File: components/entitlement/org.wso2.carbon.identity.entitlement/src/test/java/org/wso2/carbon/identity/entitlement/persistence/PAPStatusDataHandlerTest.java (hunk 3 above)
Comment (UdeshAthukorala) [reply components/entitlement/org.wso2.carbon.identity.entitlement/src/test/java/org/wso2/carbon/identity/entitlement/persistence/PAPStatusDataHandlerTest.java]: Addressed with https://github.com/wso2/carbon-identity-framework/compare/e393b967246d88953a960b77a0c4a7c355fd9dfd..f26be27fb80abcc3adb9aace8db0909c0eb3c17d

File: components/entitlement/org.wso2.carbon.identity.entitlement/src/test/java/org/wso2/carbon/identity/entitlement/persistence/PAPStatusDataHandlerTest.java (hunk 1 above)
Comment (UdeshAthukorala) [reply components/entitlement/org.wso2.carbon.identity.entitlement/src/test/java/org/wso2/carbon/identity/entitlement/persistence/PAPStatusDataHandlerTest.java]: Addressed with https://github.com/wso2/carbon-identity-framework/compare/e393b967246d88953a960b77a0c4a7c355fd9dfd..f26be27fb80abcc3adb9aace8db0909c0eb3c17d

**Review by darshanasbg — APPROVED** [review]

**Pull request merged**